### PR TITLE
Http and naked forwarding

### PR DESCRIPTION
**Pull request by tommysullivan**
main change here is to add middleware which redirects production load balanced http requests to https without forcing local dev server or pod server to redirect to https, since those environments naturally run http.

also broke up the restguru server code into smaller files

will look at adding the naked -> www redirect either in this PR or next one

oh and I added some logging middleware so we can see what requests are coming in and how those end up

### CODE REVIEW

File: docs/typescript.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+paths to learning, inluding examples, walkthrus and overviews.
+
+## Advanced Types
+
+[Advanced Types](https://www.typescriptlang.org/docs/handbook/advanced-types.html) cover union, intersection,
+mapped types, types involving key ranges, and more, built into the language as primitives!
+
+## Very Advanced Types
+
+[Typescript Toolbelt](https://www.typescriptlang.org/docs/handbook/advanced-types.html) is one of many
```
Comment (bob-bins) [inline docs/typescript.md]: did you mean
```suggestion
[Typescript Toolbelt](https://github.com/pirix-gh/ts-toolbelt) is one of many
```

File: src/utils/url/Url.ts [hunk 2]
```diff
@@ -1,12 +1,25 @@
...
+    get port() { return Number(nativeURL.port) },
+    setPath: (newPath) => copyMutateAndWrap(nativeURL, u => u.pathname = newPath),
+    setProtocol: (newProtocol) => copyMutateAndWrap(nativeURL, u => u.protocol = newProtocol),
+    setHost: (newHost) => copyMutateAndWrap(nativeURL, u => u.host = newHost),
+    toString() { return this.asString },
+    setPort(newPort) {
+        return copyMutateAndWrap(
+            nativeURL, 
+            u => u.port = newPort == defaultPort(this.protocol) 
+                ? ""
```
Comment (bob-bins) [inline src/utils/url/Url.ts]: is there a non-cosmetic reason for setting port to empty string if newPort is the default web port? and if so, would it be better to set it to None/undefined instead

File: src/testing/thenable.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { expect } from "chai";
+import { Provider } from "../utils/axioms/Provider";
+import { Consumer } from "../utils/axioms/Consumer";
+export const thenable = (prexixWord: string) => <T>(resultProvider: Provider<T>) => ({
```
Comment (bob-bins) [inline src/testing/thenable.ts]: did you mean `prefixWord`

File: src/utils/url/Url.ts (hunk 2 above)
Comment (tommysullivan) [reply src/utils/url/Url.ts]: It’s a built in type, URL, and it requires a string; setting to undefined yielded a compile error if i recall correctly.

Luckily we can mask this ugliness in our IUrl wrapper around this native api

Comment (tommysullivan) [reply src/utils/url/Url.ts]: So for example it was adding an explicit :80 or explicit 443 when the load balancer redirected in the unit tests; so I realized I wanted to eliminate that behavior, so I wanted to check, if it is the default port for the given protocol, clear the value, yielding a passing test; but to clear I had to set empty string; else couldn’t compile 🙂

Comment (tommysullivan) [reply src/utils/url/Url.ts]: So for example it was adding an explicit :80 or explicit 443 when the load balancer redirected in the unit tests; so I realized I wanted to eliminate that behavior, so I wanted to check, if it is the default port for the given protocol, clear the value, yielding a passing test; but to clear I had to set empty string; else couldn’t compile 🙂

File: src/utils/url/Url.ts [hunk 4]
```diff
@@ -1,12 +1,25 @@
...
-    asString: urlString,
-    setPath(newPath) { 
-        const tempUrl = new URL(urlString);
-        tempUrl.pathname = newPath
-        return Url(tempUrl.toString())
-    },
-    toString: () => urlString,
+export const Url = (nativeURL:URL):IUrl => ({
+    get path() { return nativeURL.pathname },
+    get host() { return nativeURL.host },
```
Comment (bob-bins) [inline src/utils/url/Url.ts]: would `fullyQualifiedDomainName` be a better name for us to switch to?

Comment (tommysullivan) [reply src/utils/url/Url.ts]: I think so!

File: docs/typescript.md (hunk 1 above)
Comment (tommysullivan) [reply docs/typescript.md]: yup good catch!

File: src/testing/thenable.ts (hunk 3 above)
Comment (tommysullivan) [reply src/testing/thenable.ts]: yup! fixed good catch

**Pull request merged**